authorizations/authorization-ranger/src/main/java/org/apache/gravitino/authorization/ranger/RangerAuthorizationPlugin.java

```diff
@@ -0,0 +1,1006 @@
+        throw new RuntimeException(e);
+      }
+    } else {
+      LOG.warn(
+          "Cannot find the Ranger policy({}) for the Gravitino securable object({})!",
+          roleName,
+          change.getSecurableObject().fullName());
+      // Don't throw exception or return false, because need support immutable operation.
+    }
+    return true;
+  }
+
+  /**
+   * 1. Find the policy base securable object. <br>
+   * 2. If the policy is exist, then user new securable object's privilege to update. <br>
+   * 3. If the policy is not exist return false. <br>
+   */
+  private boolean doUpdateSecurableObject(
+      String roleName, RoleChange.UpdateSecurableObject change) {
+    RangerPolicy policy = findManagedPolicy(change.getSecurableObject());
+
+    if (policy != null) {
+      removePolicyItem(policy, roleName, change.getSecurableObject());
+      addPolicyItem(policy, roleName, change.getNewSecurableObject());
+      try {
+        if (policy.getId() == null) {
+          rangerClient.createPolicy(policy);
+        } else {
+          rangerClient.updatePolicy(policy.getId(), policy);
+        }
+      } catch (RangerServiceException e) {
+        throw new RuntimeException(e);
+      }
+    } else {
+      LOG.warn(
+          "Cannot find the policy({}) for the securable object({})!",
+          roleName,
+          change.getSecurableObject().fullName());
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Add policy item access items base the securable object's privileges. <br>
+   * We didn't clean the policy items, because one Ranger policy maybe contain multiple Gravitino
+   * securable objects. <br>
+   */
+  private void addPolicyItem(
+      RangerPolicy policy, String roleName, SecurableObject securableObject) {
+    // First check the privilege if support in the Ranger Hive
+    checkSecurableObject(securableObject);
+
+    // Add the policy items by the securable object's privileges
+    securableObject
+        .privileges()
+        .forEach(
+            gravitinoPrivilege -> {
+              // Translate the Gravitino privilege to map Ranger privilege
+              translatePrivilege(gravitinoPrivilege.name())
+                  .forEach(
+                      mappedPrivilege -> {
+                        // Find the policy item that matches Gravitino privilege
+                        List<RangerPolicy.RangerPolicyItem> matchPolicyItems =
+                            policy.getPolicyItems().stream()
+                                .filter(
+                                    policyItem -> {
+                                      return policyItem.getAccesses().stream()
+                                          .anyMatch(
+                                              access -> access.getType().equals(mappedPrivilege));
+                                    })
+                                .collect(Collectors.toList());
+
```

Comment:
Should you check `getDenyPolicyItem`, too?

Reply:
Yes, We need to process Deny and Allow privilege.
But this PR is very big, I think I can finish this issue in the next E2E integration test PR.
I already finish E2E IT https://github.com/xunliu/gravitino/blob/issue-3963-4460/integration-test/src/test/java/org/apache/gravitino/integration/test/authorization/ranger/RangerHiveAuthorizationIT.java
waiting pull request.

Reply:
You need add a `TODO` at least.